File: php/src/Modules/Discussion/Thread.php
<?php

declare(strict_types=1);

namespace MintWiki\Discussion;

/**
 * 문서에 대한 토론 스레드를 표현하는 도메인 모델 (태스크 0410).
 *
 * Python `DiscussionThread`(src/modules/discussion/thread.py)에 대응한다.
 * id/documentId/title/createdBy/createdAt은 생성 후 바뀌지 않지만,
 * status/closedAt/pausedAt은 close()/reopen()/pause()로 바뀌므로 이
 * 모델 전체를 불변으로 두지 않는다 — User/Decision과 달리 mutable 하다.
 *
 * status는 "open"(기본값)/"closed"/"paused" 값을 갖는 평범한 문자열이며,
 * ThreadState enum으로 검증되지 않는다 — thread.py도 ThreadState를
 * 참조하지 않는다(src/modules/discussion/manifest.json 계약 노트).
 * close()/reopen()/pause()는 현재 상태와 무관하게 항상 성공하는 무조건
 * 전이이며, 잘못된 전이라는 개념이 없다. closedAt/pausedAt은 서로
 * 배타적이지 않다 — close()는 pausedAt을 리셋하지 않고, pause()도
 * closedAt을 리셋하지 않는다. reopen()만 closedAt을 null로 지운다.
 * isClosed()는 없다 — closed 여부는 "!isOpen() && !isPaused()"로
 * 판단해야 한다(같은 계약 노트).
 */
final class Thread
{
    private readonly string $title;
    private string $status;
    private ?\DateTimeImmutable $closedAt;
    private ?\DateTimeImmutable $pausedAt;

    public function __construct(
        private readonly string $id,
        private readonly string $documentId,
        string $title,
        private readonly string $createdBy,
        private readonly \DateTimeImmutable $createdAt,
        string $status = 'open',
        ?\DateTimeImmutable $closedAt = null,
        ?\DateTimeImmutable $pausedAt = null
    ) {
        if (trim($id) === '') {
            throw new EmptyThreadIdError('스레드 id는 비어있을 수 없습니다');
        }
        if (trim($documentId) === '') {
            throw new EmptyThreadDocumentIdError('문서 id는 비어있을 수 없습니다');
        }
        if (trim($title) === '') {
            throw new EmptyThreadTitleError('스레드 제목은 비어있을 수 없습니다');
        }
        if (trim($createdBy) === '') {
            throw new EmptyThreadCreatedByError('작성자 id는 비어있을 수 없습니다');
        }

        $this->title = trim(preg_replace('/\s+/u', ' ', $title));
        $this->status = $status;
        $this->closedAt = $closedAt;
        $this->pausedAt = $pausedAt;
    }

    public function id(): string
    {
        return $this->id;
    }

    public function documentId(): string
    {
        return $this->documentId;
    }

    public function title(): string
    {
        return $this->title;
    }

    public function createdBy(): string
    {
        return $this->createdBy;
    }

    public function createdAt(): \DateTimeImmutable
    {
        return $this->createdAt;
    }

    public function status(): string
    {
        return $this->status;
    }

    public function closedAt(): ?\DateTimeImmutable
    {
        return $this->closedAt;
    }

    public function pausedAt(): ?\DateTimeImmutable
    {
        return $this->pausedAt;
    }

    public function isOpen(): bool
    {
        return $this->status === 'open';
    }

    public function isPaused(): bool
    {
        return $this->status === 'paused';
    }

    public function close(\DateTimeImmutable $now): void
    {
        $this->status = 'closed';
        $this->closedAt = $now;
    }

    public function reopen(): void
    {
        $this->status = 'open';
        $this->closedAt = null;
    }

    public function pause(\DateTimeImmutable $now): void
    {
        $this->status = 'paused';
        $this->pausedAt = $now;
    }
}

File: src/modules/discussion/thread.py
"""토론 스레드 도메인 모델."""
from datetime import datetime
from typing import Optional


class EmptyThreadIdError(Exception):
    """스레드 id가 비어있을 때 발생."""

    pass


class EmptyThreadDocumentIdError(Exception):
    """스레드가 속한 문서 id가 비어있을 때 발생."""

    pass


class EmptyThreadTitleError(Exception):
    """스레드 제목이 비어있을 때 발생."""

    pass


class EmptyThreadCreatedByError(Exception):
    """스레드 작성자 id가 비어있을 때 발생."""

    pass


class DiscussionThread:
    """
    문서에 대한 토론 스레드를 나타내는 도메인 모델.

    스레드는 특정 문서를 주제로 열리며, 댓글과 상태(state)를 가진다.
    댓글 모델과 상태 열거형은 후속 태스크에서 추가되므로, 이 모델은
    상태를 단순 문자열("open"/"closed")로만 표현한다.
    """

    def __init__(
        self,
        id: str,
        document_id: str,
        title: str,
        created_by: str,
        created_at: datetime,
        status: str = "open",
        closed_at: Optional[datetime] = None,
    ):
        """
        토론 스레드를 생성한다.

        Args:
            id: 스레드의 고유 식별자
            document_id: 스레드가 속한 문서의 id
            title: 스레드 제목
            created_by: 스레드를 생성한 사용자의 id
            created_at: 스레드가 생성된 시각
            status: 스레드 상태 (기본값 "open")
            closed_at: 스레드가 닫힌 시각 (선택사항, 닫히지 않았으면 None)

        Raises:
            EmptyThreadIdError: 스레드 id가 비어있거나 공백만 있는 경우
            EmptyThreadDocumentIdError: 문서 id가 비어있거나 공백만 있는 경우
            EmptyThreadTitleError: 제목이 비어있거나 공백만 있는 경우
            EmptyThreadCreatedByError: 작성자 id가 비어있거나 공백만 있는 경우
        """
        if not id or not id.strip():
            raise EmptyThreadIdError("스레드 id는 비어있을 수 없습니다")
        if not document_id or not document_id.strip():
            raise EmptyThreadDocumentIdError("문서 id는 비어있을 수 없습니다")
        if not title or not title.strip():
            raise EmptyThreadTitleError("스레드 제목은 비어있을 수 없습니다")
        if not created_by or not created_by.strip():
            raise EmptyThreadCreatedByError("작성자 id는 비어있을 수 없습니다")

        self.id = id
        self.document_id = document_id
        self.title = title
        self.created_by = created_by
        self.created_at = created_at
        self.status = status
        self.closed_at = closed_at

    def is_open(self) -> bool:
        """스레드가 열려 있는지 확인한다."""
        return self.status == "open"

    def close(self, now: datetime) -> None:
        """스레드를 닫는다."""
        self.status = "closed"
        self.closed_at = now

    def reopen(self) -> None:
        """스레드를 다시 연다."""
        self.status = "open"
        self.closed_at = None
